### ISSUE TITLE
Lazily Evaluate Variables/threads/stacktrace for better stepping performance

### ISSUE THREAD

**Issue by JustinGrote** [PowerShell/PowerShellEditorServices#2189]
https://github.com/PowerShell/PowerShellEditorServices/blob/bbf627b597011a8c43379abe025d60df15449087/src/PowerShellEditorServices/Services/DebugAdapter/DebugService.cs#L985-L986

This line "blocks" stepping for about half a second, and I don't currently see any reason why it has to. We should return to the user as fast as possible that we have reached our breakpoint, and do this heavy processing at the point the DAP makes the `stackTrace`, `scopes`, and `variables` requests, because those are non-blocking to the UI

Assignees: 
Labels: Enhancement
Milestone: 
Projects:

**Comment by JustinGrote** [comment]
So my testing was correct and I was able to drastically speed up the UI responsiveness by returning an initial stack frame based on the invocation info, and then the subsequent DelayedStackTraceLoad awaits a task that returns the rest from the `Get-PSCallStack` invocation.

https://github.com/user-attachments/assets/f273d547-11b5-4f6c-be6e-82566482f34f

I will have to unwind some of the internal relationship between stacktrace and their variables, because per the DAP spec variables can also be lazy resolved, but PSES currently requires vars to be fully resolved to deliver a stacktrace object, so I'll more loosely couple those relationships and then put the PR up.

**Comment by JustinGrote** [comment]
Looks like this is confirmed by https://github.com/microsoft/debug-adapter-protocol/pull/344 and https://github.com/microsoft/debug-adapter-protocol/issues/79

**Comment by JustinGrote** [comment]
This is working very well but the catch is that I can't overwrite stacktraces ([asked for DAP to clarify](https://github.com/microsoft/debug-adapter-protocol/issues/508)), so I create a special `<breakpoint>` frame using similar invocation info to what @SeeminglyScience was doing in augmenting the first stack trace frame. I don't think this is too distracting or inaccurate, and it makes the experience much better. 

https://github.com/user-attachments/assets/259991b7-4699-4fb3-8989-9113941ce033


@andyleejordan if you're OK with this approach I'll rewire up the variable handling and then provide a PR.

**Comment by JustinGrote** [comment]
Out of curiosity, I also looked into how `Get-PSCallStack` works, and then found the API it uses is public.

I understand that the reasoning of the "glorious hack" as @andyleejordan put it was to unify both remote and local runspaces to serialize the info in a single channel, but if we are on a local runspace, using the API doesn't require it go thru the pipeline and completes pretty much immediately.
```csharp
runspaceContext.CurrentRunspace.Runspace.Debugger.GetCallStack()
    .Select(StackFrameDetails.Create)
```

So I am going to do a version that utilizes this if it is a local runspace or otherwise our call stack is accessible via the debugger API rather than fetching it inline at the debug prompt.

**Comment by JustinGrote** [comment]
I was also able to use the Position info to make the breakpoints more specific rather than simply at the line level
![image](https://github.com/user-attachments/assets/8b5b671d-f5b2-4902-a4a3-22295cea67a1)

**Comment by JustinGrote** [comment]
There is a `label` presentation style for stack frames specifically for this kind of artificial presentation, that looks much better:
![image](https://github.com/user-attachments/assets/5ccd3cb5-2551-4b9f-a079-eb0e97eefd57)

**Comment by andyleejordan** [comment]
This is amazing work Justin. That "glorious hack" was Rob's, for sure. I'm _personally_ of the opinion that supporting remote debugging through nested `PSSession`s is not the highest priority especially if it gets in the way of making local debugging a superb experience, because we can (and should) just be able to use that across VS Code's native remote debugging (via SSH etc.).

**Comment by JustinGrote** [comment]
@andyleejordan correct, future PRs beyond #2190 will improve that, I just didn't want to drop a giant PR on you that changes everything and takes forever to review.

**Comment by andyleejordan** [comment]
Resolved in v4.1.0 (debugger optimizations) — stack frames/variables are fetched lazily and `StackTraceHandler` short-circuits the initial UI fetch. Closing as completed.

---
_Drafted by GitHub Copilot (Claude Opus 4.8) at @andyleejordan's request._
